controllers/application_controller.go

```diff
@@ -0,0 +1,349 @@
+		argocdApp = argocd.Application()
+		err := r.client.Get(ctx, client.ObjectKey{Namespace: r.config.ArgoCD.Namespace, Name: tenantApp.GetName()}, argocdApp)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		if apierrors.IsNotFound(err) {
+			argocdApp = nil
+		}
+
+		if tenantApp.GetDeletionTimestamp() != nil {
+			res, err := r.finalize(ctx, argocdApp, tenantApp)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to finalize: %w", err)
+			}
+			return res, nil
+		}
+	}
+
+	err := r.reconcileApplication(ctx, argocdApp, tenantApp)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ApplicationReconciler) finalize(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	if !controllerutil.ContainsFinalizer(tenantApp, constants.Finalizer) &&
+		!controllerutil.ContainsFinalizer(tenantApp, argocd.ResourcesFinalizer) {
+		return ctrl.Result{}, nil
+	}
+	if argocdApp == nil {
+		controllerutil.RemoveFinalizer(tenantApp, constants.Finalizer)
+		controllerutil.RemoveFinalizer(tenantApp, argocd.ResourcesFinalizer)
+		err := r.client.Update(ctx, tenantApp)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Info("finished finalization")
+		return ctrl.Result{}, nil
+	}
+	if argocdApp.GetDeletionTimestamp() != nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	logger.Info("starting finalization")
+	err := r.client.Delete(ctx, argocdApp)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{Requeue: true}, nil
+}
+
+func (r *ApplicationReconciler) reconcileApplication(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) error {
+	logger := log.FromContext(ctx)
+
+	removed, err := r.fixProject(ctx, argocdApp, tenantApp)
+	if err != nil {
+		logger.Error(err, "failed to validate application project")
+		return err
+	}
+	if removed {
+		return nil
+	}
+
+	// Sync application spec from tenant to argocd.
+	err = r.syncApplicationSpec(ctx, argocdApp, tenantApp)
+	if err != nil {
+		logger.Error(err, "failed to sync application spec")
+		return err
+	}
+	// Sync application status from argocd to tenant.
+	err = r.syncApplicationStatus(ctx, argocdApp, tenantApp)
+	if err != nil {
+		logger.Error(err, "failed to sync application status")
+		return err
+	}
+	return nil
+}
+
+func (r *ApplicationReconciler) fixProject(ctx context.Context, argocdApp *unstructured.Unstructured, tenantApp *unstructured.Unstructured) (removed bool, err error) {
+	logger := log.FromContext(ctx)
+
+	ns := &corev1.Namespace{}
+	err = r.client.Get(ctx, client.ObjectKey{Name: tenantApp.GetNamespace()}, ns)
+	if err != nil {
+		return
+	}
+	tenantName := ns.Labels[constants.OwnerTenant]
+	if tenantName == "" {
+		if argocdApp != nil && argocdApp.GetDeletionTimestamp() == nil {
+			logger.Info("Remove unmanaged application")
```

Comment:
It was difficult to understand at first because it was hard to associate the word "unmanged" with the fact that the NS where the application was deployed was not labeled cattage.cybozu.io/tenant.
I don't have any other ideas on how to express properly, so this is just a comment. Sorry.

Reply:
Described that in the comment.